miden-lib/asm/sat/epilogue.masm

```diff
@@ -0,0 +1,229 @@
+    push.0
+
+    # prepare address and stack for reading assets
+    movup.2 push.CREATED_NOTE_ASSETS_OFFSET add padw padw padw
+
+    # check if the number of assets is greater then 0
+    dup.14 dup.14 neq
+
+    # loop and read assets from memory
+    while.true
+        # read assets from memory.
+        # if this is the last permutation of the loop and we have an odd number of assets then we
+        # implicitly pad the last word of the hasher rate with ZERO's by reading from empty memory.
+        mem_stream
+
+        # check if we should loop again
+        movup.13 push.2 add dup movdn.14 dup.15 neq
+    end
+
+    # extract digest from hasher rate elements (h_0, ..., h_3)
+    dropw swapw dropw
+
+    # drop accessory variables from stack
+    movup.4 drop
+    movup.4 drop
+    movup.4 drop
+
+    # save vault hash to memory
+    movup.4 push.CREATED_NOTE_VAULT_HASH_OFFSET add mem_storew
+end
+
+#! Computes the hash of a created note with index i. This is computed as follows:
+#! - we define, recipient =
+#!       hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash)
+#! - we then compute the created note hash as:
+#!       hash(recipient, vault_hash)
+#!
+#! Stack: [i]
+#! Output: [CREATED_NOTE_HASH]
+#!
+#! - i is the index of the created note to hash
+#! - CREATED_NOTE_HASH is the hash of the created note with index i
+proc.compute_created_note_hash
+    # pad capacity elements of hasher
+    padw
+
+    # insert created note recipient into the first four elements of the hasher rate
+    padw dup.8 exec.utils::created_note_data_ptr push.CREATED_NOTE_RECIPIENT_OFFSET add mem_loadw
+
+    # populate the last four elements of the hasher rate with the created note vault hash
+    movup.8 exec.compute_created_note_vault_hash
+
+    # compute crated note hash and extract digest
+    hperm dropw swapw dropw
+end
+
+#! Computes a commitment to the created notes. This is computed as a sequential hash of
+#! (note_hash, note_metadata) tuples.
+#!
+#! Stack: []
+#! Output: [CREATED_NOTES_COMMITMENT]
+#!
+#! - CREATED_NOTES_COMMITMENT is the commitment of the created notes.
+proc.process_created_notes
+    # get the number of created notes from memory
+    push.NUM_CREATED_NOTES_PTR mem_load
+
+    # initialize counter for created note hashing
+    push.0
+
+    # prepare stack for hashing
+    padw padw padw
+
+    # check if the number of created notes is greater then 0. Conditional for the while loop.
+    dup.13 dup.13 neq
+
+    # loop and hash created notes
+    while.true
+        # clear hasher rate
+        dropw
+
+        # populate hasher rate with created note
+        dup.8 exec.compute_created_note_hash swapw
+
+        # populate hasher rate with second created note
+        dup.12 exec.utils::created_note_data_ptr mem_loadw
+
+        # execute permutation
+        hperm
+
+        # increment created note counter and check if we should loop again
+        movup.12 push.1 add dup movdn.13 dup.14 neq
+
+    end
```

Comment:
At the minute we don't persist the `note_hash` but we could introduce it as the `0` element such that we have:
```
const.CREATED_NOTE_HASH_OFFSET=0
const.CREATED_NOTE_META_OFFSET=1
const.CREATED_NOTE_RECIPIENT_OFFSET=2
const.CREATED_NOTE_VAULT_HASH_OFFSET=3
const.CREATED_NOTE_NUM_ASSETS_OFFSET=4
const.CREATED_NOTE_ASSETS_OFFSET=5
```

Reply:
I was thinking more like if the note hash+meta were next to each other for all notes, it could make things simpler. i.e., something like this:
```
mem[x+0] = note1_hash
mem[x+1] = note1_meta
mem[x+2] = note2_hash
mem[x+3] = note2_meta
mem[x+4] = note3_hash
mem[x+5] = note3_meta
etc.
```
This would require storing note hash and meta in memory region which is not adjacent to all of created note data - so, might introduce extra complexities. But we should understand if this could result in noticeable savings. Probably not for this PR though.

Reply:
So I've gone ahead and updated it to use the pattern I've described.  Now instead of using a raw counter variable to terminate the loop we use a memory address.  I think what would be good is if we could modify the semantics of `mem_stream` such that we can specify an immediate value such as `mem_stream.1024`.  This will imply that we should increment the memory address by 1024.  This flexibility could give us more freedom in terms of how we layout memory.  At the minute it assumes that data is layed out sequentially.

Reply:
> I think what would be good is if we could modify the semantics of `mem_stream` such that we can specify an immediate value such as `mem_stream.1024`.

Interesting! Let's create an issue in `miden-vm` repo for this.